### PR TITLE
[Booking]:  added CRUD operations for Booking

### PR DESCRIPTION
**Pull request by DPotapov93**

### CODE REVIEW

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/bookings")
+@Tag(name = "Booking management", description = "Endpoints for managing bookings")
+@RequiredArgsConstructor
+public class BookingController {
+    private final BookingService bookingService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:37]: It's better to use `@PreAuthorize("hasRole('USER')")` here to have code consistency in project.
Also do the same changes to other `@PreAuthorize` annotations

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    @ResponseStatus(HttpStatus.CREATED)
+    public BookingResponseDto createBooking(
+            @RequestBody @Valid CreateBookingRequestDto requestDto,
+            Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return bookingService.createBooking(user.getId(), requestDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @GetMapping()
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:49]:
```suggestion
    @GetMapping
```

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            description = "Retrieves bookings based on user ID and their status"
+    )
+    public List<BookingResponseDto> getBookingsByUserIdAndStatus(
+            @RequestParam(name = "user_id") Optional<Long> optionalId,
+            @RequestParam(name = "status") Booking.Status status,
+            Pageable pageable
+    ) {
+        return optionalId.map(id ->
+                        bookingService.getBookingsByUserIdAndStatus(id, status, pageable))
+                .orElseGet(() -> bookingService.getBookingsByStatus(status, pageable));
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:61]: Looks too complicated for Controller. My suggestion is to make this check in the service layer.

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/my")
+    @Operation(summary = "Retrieves user bookings")
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:66]: You forgot to add description here

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        User user = (User) authentication.getPrincipal();
+        return bookingService.getBookingHistory(user.getId(), pageable);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(
+            summary = "Get the booking",
+            description = "Provides information about a specific booking"
+    )
+    @GetMapping("/{id}")
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:80]: More consistent variant
```suggestion
    @PreAuthorize("hasRole('USER')")
    @GetMapping("/{id}")
    @Operation(
            summary = "Get the booking",
            description = "Provides information about a specific booking"
    )
```

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        return bookingService.updateBookingDetails(user.getId(), bookingId, requestDto);
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(
+            summary = "Delete the booking",
+            description = "Enables the cancellation of bookings"
+    )
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:106]: More consistent variant:
```suggestion
    @PreAuthorize("hasRole('USER')")
    @DeleteMapping("/{id}")
    @Operation(
            summary = "Delete the booking",
            description = "Enables the cancellation of bookings"
    )
    @ResponseStatus(HttpStatus.NO_CONTENT)
```

File: src/main/java/com/example/accommodationbookingapp/controller/BookingController.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    public void deleteBooking(@PathVariable Long id, Authentication authentication) {
```
Comment (vadymhrnk) [inline src/main/java/com/example/accommodationbookingapp/controller/BookingController.java:107]: I don't see that you use `Authentication authentication` in method body. Consider changing this

**Review by vadymhrnk — CHANGES_REQUESTED** [review]
Good job! Left some suggestions for you to consider.

**Review by vadymhrnk — APPROVED** [review]
Nicely done. Looks good to me.

**Review by MinoDentori — APPROVED** [review]
GJ!

**Pull request merged**